src/utils/rpc_selector.rs

```diff
@@ -0,0 +1,373 @@
+use std::sync::{atomic::AtomicUsize, Arc, PoisonError};
+
+use eyre::Result;
+use rand::distr::weighted::WeightedIndex;
+use rand::prelude::*;
+
+use crate::config::RpcConfig;
+
+#[derive(Debug)]
+pub enum RpcSelectorError {
+    NoProviders,
+    MutexLockError(String),
```

Comment:
What happens in case of mutex lock error?
Is it even possible because we use non cached instances?

In any case we must ensure that this error is not propagated and that we would try to retry to obtain lock.

Should we use locks at all?